trunner/device.py

```diff
@@ -30,13 +33,203 @@
+        self.proc = None
+
+    def read_output(self):
+        if is_github_actions():
+            logging.info('::group::psu\n')
+
+        while True:
+            line = self.proc.stdout.readline().decode('utf-8')
+            if not line:
+                break
+
+            logging.info(line)
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        self.proc = subprocess.Popen(
+            ['psu', f'{self.script}'],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.DEVNULL,
+            cwd=self.cwd
+        )
+
+        self.read_output()
+        self.proc.wait()
+        if self.proc.returncode != 0:
+            logging.error(f'Command {" ".join(self.proc.args)} with pid {self.proc.pid} failed!\n')
+            raise Exception('Flashing IMXRT106x failed\n')
+
+
+class Phoenixd:
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        port,
+        baudrate=460800,
+        dir='.',
+        cwd=_BOOT_DIR,
+        wait_dispatcher=True
+    ):
+        self.port = port
+        self.baudrate = baudrate
+        self.dir = dir
+        self.cwd = cwd
+        self.proc = None
+        self.reader_thread = None
+        self.reader_lock = None
+        self.wait_dispatcher = wait_dispatcher
+        self.output_buffer = ''
+
+    def _reader(self):
+        while True:
+            line = self.proc.readline()
+            if not line:
+                break
+
+            if self.wait_dispatcher:
+                msg = f'Starting message dispatcher on [{self.port}] (speed={self.baudrate})'
+                if msg in line:
+                    self.wait_dispatcher = False
+                    self.reader_lock.release()
+
+            self.output_buffer += line
+
+    def run(self):
+        # Use pexpect.spawn to run process as PTY, so it will flush at new line
+        self.proc = pexpect.spawn(
+            'phoenixd',
+            ['-p', self.port,
+             '-b', str(self.baudrate),
+             '-s', self.dir],
+            cwd=self.cwd,
+            encoding='utf-8'
+        )
+
+        if self.wait_dispatcher:
+            self.reader_lock = threading.Lock()
+            self.reader_lock.acquire()
+
+        self.reader_thread = threading.Thread(target=self._reader)
+        self.reader_thread.start()
+
+        if self.wait_dispatcher:
+            # When dispatcher starts then reader will release lock
+            self.reader_lock.acquire(timeout=10)
+
+        return self.proc
+
+    def output(self):
+        output = self.output_buffer
+        if is_github_actions():
+            output = '::group::Pheonixd\n' + output + '::endgroup::\n'
+
+        return output
+
+    def kill(self):
+        os.killpg(os.getpgid(self.proc.pid), signal.SIGTERM)
+        self.reader_thread.join()
+        if self.reader_lock.locked():
+            self.reader_lock.release()
+
+    def __enter__(self):
+        self.run()
```

Comment:
If exception is raised inside run function after invoking spawn than kill method will not be called (exception in \__enter\__  does not result in \__exit\__ call).

Reply:
Do you suggest that I should put threading code into `try: except:`? I also changed a little bit `Lock` handling.

Reply:
I was mostly concerned about timeout exception from Lock.acquire. This is not actually a case as acquire simply returns false on timeout. 

It is fine now - can be merged as is.

### for future reference: 

Generic solution to deal with this case would be to use:
```python
def __enter__(self):
    try:
        self.run()
    except Exception:
        self.__exit__(None, None, None) # if exit uses args than None should be changed to proper values
```
Note that `__exit__` in this case needs to do additional checks (ie. self.proc could be None if exception occurred before spawn)

My preferred way of implementing context managers is to use [contextlib.contextmanager decorator](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) as it allows using try and with statements in standard way.

```python
class Phoenixd:
    @contextlib.contextmanager
    def running(self):
        try:
            self.run()
            yield self # all code inside with statement happens here
        except Exception:
            self.kill()
```

and than to use it:

```python
pxd = Phoenixd(...)
with pxd.running():
    # ...
```